cmake/cc_rules.cmake

```diff
@@ -39,8 +40,16 @@ function(cc_library name)
     ${FIREBASE_BINARY_DIR}
     ${FIREBASE_SOURCE_DIR}
   )
+
   target_link_libraries(${name} PUBLIC ${ccl_DEPENDS})
 
+  foreach(external_dep ${ccl_EXTERNAL_DEPENDS})
+    target_link_libraries(${name} PUBLIC ${external_dep})
+    get_target_property(external_headers ${external_dep} INTERFACE_INCLUDE_DIRECTORIES)
+    # Ignore any warnings from external headers.
+    target_include_directories(${name} SYSTEM PUBLIC ${external_headers})
```

Comment:
This is a pretty indirect way of handling this problem and it's suboptimal because it seems like everything that indirectly depends on grpc++ needs to add an EXTERNAL_DEPENDS on it to avoid also breaking.

Ideally we'd fix the warnings to prevent this problem in the first place but I realize that's not terribly practical. A better path would likely be to pass `-Wno-error` using `target_compile_options`.

A PUBLIC target_compile_option would basically work as-is, but makes this configuration value viral. Everything above remote in our system would end up compiled this way.

You can contain the blast radius of this by making this a a PRIVATE compile option and then ensure that the gRPC headers containing the warnings are never included from any of our headers--only implementation files. That may not be practical though and the PUBLIC option isn't unreasonable. Ultimately we're going to have to disable warnings-as-errors anyway. It's extremely unfriendly.

Reply:
I certainly don't feel strongly about this particular approach, but I'd like to clarify:
* what do you mean by indirect? The basic idea is ignoring warnings from third-party dependencies, similar to `-isystem` flag (which CMake probably emulates with their `SYSTEM`). Even if we didn't have `-Werror`, omitting warnings not from our code seems like a good thing. Perhaps the name doesn't reflect the intention well enough?
* the transitive dependency works well, AFAICS: currently, `remote_test` target successfully depends on `remote` without depending on `grpc++`. For the sake of experiment, I tried this:
```cmake
cc_library(
  firebase_firestore_remote_grpc
  SOURCES
    grpc_call.h
    grpc_completion.h
    grpc_completion.cc
    grpc_connection.h
    grpc_connection.cc
    grpc_root_certificate_finder.h
    grpc_root_certificate_finder_generated.cc
    grpc_root_certificates_generated.h
    grpc_root_certificates_generated.cc
    grpc_stream.h
    grpc_stream.cc
    grpc_stream_observer.h
    grpc_streaming_reader.h
    grpc_streaming_reader.cc
    grpc_unary_call.h
    grpc_unary_call.cc
    grpc_util.h
    grpc_util.cc
    grpc_util.cc

  DEPENDS
    firebase_firestore_model
    firebase_firestore_remote_connectivity_monitor
    firebase_firestore_util
  EXTERNAL_DEPENDS
    grpc++
)

cc_library(
  firebase_firestore_remote
  SOURCES
    exponential_backoff.h
    exponential_backoff.cc
    serializer.h
    serializer.cc
  DEPENDS
    # TODO(b/111328563) Force nanopb first to work around ODR violations
    protobuf-nanopb

    firebase_firestore_model
    firebase_firestore_nanopb
    firebase_firestore_protos_nanopb
    firebase_firestore_remote_connectivity_monitor
    firebase_firestore_remote_grpc
    firebase_firestore_util
    firebase_firestore_version
)
```
i.e., `remote` doesn't depend on `grpc++` directly but instead depends on `remote_grpc`, and it still builds successfully.

> Ideally we'd fix the warnings to prevent this problem in the first place but I realize that's not terribly practical. 

The warnings are about unused named parameters; I think fixing that has a good chance of being merged, but I'd be concerned about potential for breakage each time gRPC is upgraded and might start producing more warnings (also, assuming the warnings are fixed each time, waiting for the next gRPC release).

FWIW, I'm okay with our code being `-Werror`, but would certainly prefer that not to be the case for our dependencies.

Reply:
Sorry, I thought I saw you adding dependencies on `grpc++` in places that didn't need them to preserve this behavior. Please ignore that part.

I still think `-isystem` isn't a great solution though because that flag is not principally about suppressing warnings, but instead treating the path as a system include directory. This could potentially interact poorly with various hacks the compilers have in place to fix broken system header files. In particular it affects the handling of `#include_next`. gRPC even includes headers with the same names as system headers like `<time.h>` so this seems like the wrong classification for these directories.

Other projects have [discussed this approach](https://github.com/conan-io/conan/issues/2629#issuecomment-374077886) because it is appealing, but I think we need to find something else.

We could also locally disable these warnings before including these headers, though I don't know how practical that is.

Reply:
Thanks for the link, this was a very informative read.

Actually, and surprisingly, this appears to be a non-issue. Previously, I had `grpc` library among the dependencies, which generated the warnings-errors. Actually, what I need is `grpc++` library, while `grpc` is unnecessary. Including the same headers from `grpc++` doesn't lead to warnings for some reason.